### ISSUE TITLE
systemd socket activation fails when twisted shutdown the inherited socket

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#5621]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)| @wallrj reported|
|-|-|
|Trac ID|trac#5621|
|Type|defect|
|Created|2012-04-07 01:56:48Z|
|Branch|https://github.com/twisted/twisted/tree/do-not-shutdown-sockets-5621-3|

I've been testing and documenting systemd endpoints ([[#5601](https://github.com/twisted/twisted/issues/5601)](https://github.com/twisted/twisted/issues/5601)).

I find that if I kill (SIGTERM) the twistd process, systemd complains that the listening socket has been shutdown.

It then refuses any further connections.

The test setup is described in #5601:[5601-systemd-documentation.patch](None/5601-systemd-documentation.patch)

```
Apr  7 02:45:23 zorin twistd[6409]: 2012-04-07 02:45:23+0100 [-] Log opened.
Apr  7 02:45:23 zorin twistd[6409]: 2012-04-07 02:45:23+0100 [-] twistd 12.0.0+r34099 (/usr/bin/python 2.7.2) starting up.
Apr  7 02:45:23 zorin twistd[6409]: 2012-04-07 02:45:23+0100 [-] reactor class: twisted.internet.pollreactor.PollReactor.
Apr  7 02:45:23 zorin twistd[6409]: 2012-04-07 02:45:23+0100 [-] Site starting on 8080
Apr  7 02:45:23 zorin twistd[6409]: 2012-04-07 02:45:23+0100 [-] Starting factory <twisted.web.server.Site instance at 0x19fa7e8>
Apr  7 02:45:23 zorin twistd[6409]: 2012-04-07 02:45:23+0100 [HTTPChannel,0,127.0.0.1] 127.0.0.1 - - [07/Apr/2012:01:45:23 +0000] "GET / HTTP/1.1" 200 1047 "-" "curl/7.21.7 (x86_64-redhat-linux-gnu) libcurl/7.21.7 NSS/3.13.3.0 zlib/1.2.5 libidn/1.22 libssh2/1.2.7"
Apr  7 02:45:37 zorin twistd[6409]: 2012-04-07 02:45:37+0100 [-] Received SIGTERM, shutting down.
Apr  7 02:45:37 zorin twistd[6409]: 2012-04-07 02:45:37+0100 [-] (TCP Port 8080 Closed)
Apr  7 02:45:37 zorin twistd[6409]: 2012-04-07 02:45:37+0100 [-] Stopping factory <twisted.web.server.Site instance at 0x19fa7e8>
Apr  7 02:45:37 zorin twistd[6409]: 2012-04-07 02:45:37+0100 [-] Main loop terminated.
Apr  7 02:45:37 zorin twistd[6409]: 2012-04-07 02:45:37+0100 [-] Server Shut Down.
Apr  7 02:45:37 zorin systemd[1]: twistedweb.socket: Got POLLHUP on a listening socket. The service probably invoked shutdown() on it, and should better not do that.
Apr  7 02:45:37 zorin systemd[1]: Unit twistedweb.socket entered failed state.
```

<details><summary>Searchable metadata</summary>

```
trac-id__5621 5621
type__defect defect
reporter__rwall rwall
priority__normal normal
milestone__ 
branch__branches_do_not_shutdown_sockets_5621_3 branches/do-not-shutdown-sockets-5621-3
branch_author__exarkun exarkun
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1333763808000000 1333763808000000
changetime__1334749607000000 1334749607000000
version__None None
owner__exarkun exarkun

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun commented|
|-|-|

(In [34142]) Branching to 'do-not-shutdown-sockets-5621'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun commented|
|-|-|

(In [34204]) Branching to 'do-not-shutdown-sockets-5621-2'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun commented|
|-|-|

(In [34208]) Branching to 'do-not-shutdown-sockets-5621-3'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun commented|
|-|-|

(In [34209]) Do not shutdown adopted ports.

refs [#5621](https://github.com/twisted/twisted/issues/5621)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun commented|
|-|-|

Originally I wanted to get rid of all shutdown calls.  After some discussion about which use cases (applications calling fork themselves without calling exec) would be broken by this, I decided to go for the smaller change of just not calling shutdown on adopted ports.

[Build Results](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/do-not-shutdown-sockets-5621-3)

I haven't actually tried this under systemd to see if it fixes the reported behavior.  I can if no one else has easier access.  I think it should be fine now, though, because we're not calling shutdown anymore.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_6"></a>|@therve set owner to @exarkun|
|-|-|

Looks good. I haven't tested it for real either, but it fixes the mentioned problem. Please add a NEWS fragment and merge.

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_7"></a>|@wallrj commented|
|-|-|

It works. Tested on Fedora 16.

 * Start systemd as a non root user
```
[richard@zorin branches]$ cd do-not-shutdown-sockets-5621-3/
[richard@zorin do-not-shutdown-sockets-5621-3]$ whbranch
Twisted: do-not-shutdown-sockets-5621-3
Pydoctor: trunk
Divmod: trunk
[richard@zorin do-not-shutdown-sockets-5621-3]$ systemd --user
```

 * Tell systemd to listen on port 8080
```
[richard@zorin branches]$ systemctl  --user start twistedweb.socket
[richard@zorin branches]$ netstat -lntp
(Not all processes could be identified, non-owned process info
 will not be shown, you would have to be root to see it all.)
Active Internet connections (only servers)
Proto Recv-Q Send-Q Local Address               Foreign Address             State       PID/Program name
tcp        0      0 127.0.0.1:631               0.0.0.0:*                   LISTEN      -
tcp        0      0 127.0.0.1:25                0.0.0.0:*                   LISTEN      -
tcp        0      0 0.0.0.0:8080                0.0.0.0:*                   LISTEN      12424/systemd
tcp        0      0 0.0.0.0:22                  0.0.0.0:*                   LISTEN      -
tcp        0      0 :::631                      :::*                        LISTEN      -
tcp        0      0 :::22                       :::*                        LISTEN      -
```

 * Connect to port 8080
```
[richard@zorin branches]$ printf "GET / HTTP/1.0\r\n\r\n" | nc localhost 8080 | head
HTTP/1.0 200 OK
Date: Tue, 17 Apr 2012 22:40:50 GMT
Content-Length: 1047
Content-Type: text/html; charset=utf-8
Server: TwistedWeb/12.0.0
...
```


 * Systemd activates twisted.

```
Apr 17 23:40:50 zorin twistd[12430]: 2012-04-17 23:40:50+0100 [-] Log opened.
Apr 17 23:40:50 zorin twistd[12430]: 2012-04-17 23:40:50+0100 [-] twistd 12.0.0 (/usr/bin/python 2.7.2) starting up.
Apr 17 23:40:50 zorin twistd[12430]: 2012-04-17 23:40:50+0100 [-] reactor class: twisted.internet.pollreactor.PollReactor.
Apr 17 23:40:50 zorin twistd[12430]: 2012-04-17 23:40:50+0100 [-] Site starting on 8080
Apr 17 23:40:50 zorin twistd[12430]: 2012-04-17 23:40:50+0100 [-] Starting factory <twisted.web.server.Site instance at 0x13317a0>
Apr 17 23:40:50 zorin twistd[12430]: 2012-04-17 23:40:50+0100 [HTTPChannel,0,127.0.0.1] 127.0.0.1 - - [17/Apr/2012:22:40:50 +0000] "GET / HTTP/1.0" 200 1047 "-" "-"
```

 * Kill the twisted process
```
[richard@zorin branches]$ kill 12430
```

 * Twisted closes down and systemd no longer complains that the socket was shutdown and it keeps listening.

```
Apr 17 23:41:05 zorin twistd[12430]: 2012-04-17 23:41:05+0100 [-] Received SIGTERM, shutting down.
Apr 17 23:41:05 zorin twistd[12430]: 2012-04-17 23:41:05+0100 [-] (TCP Port 8080 Closed)
Apr 17 23:41:05 zorin twistd[12430]: 2012-04-17 23:41:05+0100 [-] Stopping factory <twisted.web.server.Site instance at 0x13317a0>
Apr 17 23:41:05 zorin twistd[12430]: 2012-04-17 23:41:05+0100 [-] Main loop terminated.
Apr 17 23:41:05 zorin twistd[12430]: 2012-04-17 23:41:05+0100 [-] Server Shut Down.
```

```
[richard@zorin branches]$ netstat -lntp
(Not all processes could be identified, non-owned process info
 will not be shown, you would have to be root to see it all.)
Active Internet connections (only servers)
Proto Recv-Q Send-Q Local Address               Foreign Address             State       PID/Program name
tcp        0      0 127.0.0.1:631               0.0.0.0:*                   LISTEN      -
tcp        0      0 127.0.0.1:25                0.0.0.0:*                   LISTEN      -
tcp        0      0 0.0.0.0:8080                0.0.0.0:*                   LISTEN      12424/systemd
tcp        0      0 0.0.0.0:22                  0.0.0.0:*                   LISTEN      -
tcp        0      0 :::631                      :::*                        LISTEN      -
tcp        0      0 :::22                       :::*                        LISTEN      -
```

 * Connect to port 8080 again and systemd restarts twisted and twisted responds.
```
[richard@zorin branches]$ printf "GET / HTTP/1.0\r\n\r\n" | nc localhost 8080 | head
HTTP/1.0 200 OK
Date: Tue, 17 Apr 2012 22:41:22 GMT
Content-Length: 1047
Content-Type: text/html; charset=utf-8
Server: TwistedWeb/12.0.0
```

```
Apr 17 23:41:22 zorin twistd[12436]: 2012-04-17 23:41:22+0100 [-] Log opened.
Apr 17 23:41:22 zorin twistd[12436]: 2012-04-17 23:41:22+0100 [-] twistd 12.0.0 (/usr/bin/python 2.7.2) starting up.
Apr 17 23:41:22 zorin twistd[12436]: 2012-04-17 23:41:22+0100 [-] reactor class: twisted.internet.pollreactor.PollReactor.
Apr 17 23:41:22 zorin twistd[12436]: 2012-04-17 23:41:22+0100 [-] Site starting on 8080
Apr 17 23:41:22 zorin twistd[12436]: 2012-04-17 23:41:22+0100 [-] Starting factory <twisted.web.server.Site instance at 0x11587a0>
Apr 17 23:41:22 zorin twistd[12436]: 2012-04-17 23:41:22+0100 [HTTPChannel,0,127.0.0.1] 127.0.0.1 - - [17/Apr/2012:22:41:22 +0000] "GET / HTTP/1.0" 200 1047 "-" "-"
```

Thanks exarkun!

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_8"></a>|@exarkun commented|
|-|-|

> It works. Tested on Fedora 16.

Thanks for testing and for the thorough report!

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun set status to `closed`|
|-|-|

(In [34225]) Merge do-not-shutdown-sockets-5621-3

Author: exarkun
Reviewer: therve, rwall
Fixes: [#5621](https://github.com/twisted/twisted/issues/5621)

Remove the `shutdown(2)` call on file descriptors added to the reactor using
`IReactorSocket.adoptStreamPort`.  The socket is still closed, which will
have the same effect if and only if no other references to the socket exist
(for example, in a systemd process, where it was originally created).
